pyomo/gdp/util.py

```diff
@@ -78,22 +81,136 @@ def clone_without_expression_components(expr, substitute=None):
                                                 remove_named_expressions=True)
     return visitor.walk_expression(expr)
 
-def preprocess_targets(targets):
-    preprocessed_targets = []
+class GDPTree:
+    def __init__(self):
+        self._adjacency_list = {}
+        self._in_degrees = defaultdict(lambda: 0)
+        # This needs to be ordered so that topological sort is deterministic
+        self._vertices = []
+
+    @property
+    def vertices(self):
+        return self._vertices
+
+    def add_node(self, u):
+        if u not in self._vertices:
+            self._vertices.append(u)
+
+    def _update_in_degree(self, v):
+        self._in_degrees[v] += 1
+
+    def add_edge(self, u, v):
+        if u in self._adjacency_list:
+            self._adjacency_list[u].append(v)
+        else:
+            self._adjacency_list[u] = [v]
+        self._update_in_degree(v)
+        if u not in self._vertices:
+            self._vertices.append(u)
+        if v not in self._vertices:
+            self._vertices.append(v)
+
+    def _visit_vertex(self, u, leaf_to_root):
+        if u in self._adjacency_list:
+            for v in self._adjacency_list[u]:
+                if v not in leaf_to_root:
+                    self._visit_vertex(v, leaf_to_root)
+        # we're done--we've been to all its children
+        leaf_to_root.append(u)
+
+    def _topological_sort(self):
+        # this is reverse of the list we should return (but happens to be what
+        # we want for hull and bigm)
+        leaf_to_root = []
+        for u in self.vertices:
+            if u not in leaf_to_root:
+                self._visit_vertex(u, leaf_to_root)
+
+        return leaf_to_root
+
+    def topological_sort(self):
+        return reversed(self._topological_sort())
+
+    def reverse_topological_sort(self):
+        return self._topological_sort()
+
+    def in_degree(self, u):
+        return self._in_degrees[u]
+
+def _parent_disjunct(obj):
+    parent = obj.parent_block()
+    while parent is not None:
+        if parent.ctype is Disjunct:
+            return parent
+        parent = parent.parent_block()
+
+    return None
+
+def _gather_disjunctions(block, gdp_tree):
+    to_explore = ComponentSet([block])
```

Comment:
I am not sure that `ComponentSet` is deterministic ordered.  Is that an issue?  Can `to_explore` just be a list?